src/fiber/fiber.ml

```diff
@@ -1,243 +1,455 @@
+
+    let unset var f k = Unset_var (var, f, k)
+
+    let create () = create ~name:"var" (fun _ -> Dyn.string "var")
   end
 
-  let add_refs n =
-    let t = !current in
-    match t.on_release with
-    | Do_nothing -> ()
-    | Exec r -> r.ref_count <- r.ref_count + n
-
-  let rec forward_exn_with_backtrace t exn =
-    match t.on_error with
-    | None -> Exn_with_backtrace.reraise exn
-    | Some { ctx; run } -> (
-      current := ctx;
-      try run exn with
-      | exn -> forward_error exn)
-
-  and forward_error exn =
-    let exn = Exn_with_backtrace.capture exn in
-    forward_exn_with_backtrace !current exn
-
-  let deref t =
-    match t.on_release with
-    | Do_nothing -> ()
-    | Exec r -> (
+  type fill = Fill : 'a ivar * 'a -> fill
+
+  module Jobs = struct
+    type t =
+      | Empty
+      | Job : context * ('a -> effect) * 'a -> t
+      | Concat : t * t -> t
+
+    let concat a b =
+      match (a, b) with
+      | Empty, x
+      | x, Empty ->
+        x
+      | _ -> Concat (a, b)
+
+    let rec enqueue_readers (readers : (_, [ `Empty ]) ivar_state) x jobs =
+      match readers with
+      | Empty -> jobs
+      | Empty_with_readers (ctx, k, readers) ->
+        enqueue_readers readers x (Concat (Job (ctx, k, x), jobs))
+
+    let fill_ivar ivar x jobs =
+      match ivar.state with
+      | Full _ -> failwith "Fiber.Ivar.fill"
+      | (Empty | Empty_with_readers _) as readers ->
+        ivar.state <- Full x;
+        enqueue_readers readers x jobs
+
+    let rec exec_fills fills acc =
+      match fills with
+      | [] -> acc
+      | Fill (ivar, x) :: fills ->
+        let acc = fill_ivar ivar x acc in
+        exec_fills fills acc
+
+    let exec_fills fills = exec_fills (List.rev fills) Empty
+  end
+
+  exception Toplevel_exception of Exn_with_backtrace.t
+
+  let rec length_and_rev l len acc =
+    match l with
+    | [] -> (len, acc)
+    | x :: l -> length_and_rev l (len + 1) (x :: acc)
+
+  let length_and_rev l = length_and_rev l 0 []
+
+  let list_of_option_array =
+    let rec loop arr i acc =
+      if i = 0 then
+        acc
+      else
+        let i = i - 1 in
+        match arr.(i) with
+        | None -> assert false
+        | Some x -> loop arr i (x :: acc)
+    in
+    fun a -> loop a (Array.length a) []
+
+  let run (type a) (t : a t) ~iter =
+    let module M = struct
+      type value += X of a
+    end in
+    let rec loop : Jobs.t -> a = function
+      | Empty -> iter () |> Nonempty_list.to_list |> Jobs.exec_fills |> loop
+      | Job (ctx, run, x) -> exec ctx run x Empty
+      | Concat (a, b) -> loop2 a b
+    and loop2 a b =
+      match a with
+      | Empty -> loop b
+      | Job (ctx, run, x) -> exec ctx run x b
+      | Concat (a1, a2) -> loop2 a1 (Jobs.concat a2 b)
+    and exec : 'a. context -> ('a -> effect) -> 'a -> Jobs.t -> a =
+     fun ctx k x jobs ->
+      match k x with
+      | exception Toplevel_exception exn -> Exn_with_backtrace.reraise exn
```

Comment:
Why do we drop the backtrace here?

Reply:
It's already part of `exn` (it's a `Exn_with_backtrace.t`).